subprojects/language-java/src/main/java/org/gradle/api/tasks/compile/JavaCompile.java

```diff
@@ -408,6 +409,7 @@ public FileCollection getClasspath() {
      */
     @SkipWhenEmpty
     @IgnoreEmptyDirectories
+    @NormalizeLineEndings
```

Comment:
Outside of Java compilation, can this normalizer be used by anything else? I'm wondering if this makes sense to expose as a fully generic feature initially.

When I think about where else we could use this (e.g., code quality), it immediately breaks down. If you had a checkstyle rule that checked for `\n` line endings only, you wouldn't want to ignore line endings.

I think this also breaks for Groovy where the line endings inside a multi-line string are meaningful (for good or bad). Not sure which way Kotlin falls with it's multi-line strings. Java at the very last always uses `\n` regardless of what the source has in a text block.

I'm also not sure if this annotation can't be used by other tasks where Java source is an input (e.g., does the line endings in javadoc effect javadoc output?).

I think I would agree that most of the time the line endings _should not matter_, but this looks like it might be an easy way for people to introduce subtle bugs that make them not trust the build very much.

Reply:
It could be used by any task that produces the same output regardless of the line endings in the input.  We're only turning this on for Java because it implicitly normalizes line endings in Strings during compilation.  That's not the case for Groovy and Kotlin (as far as I'm aware) which is why we're not doing it for `SourceTask`.  

I simply don't know about javadoc.  We can always turn it on later if we find that it's useful.

> I think I would agree that most of the time the line endings should not matter, but this looks like it might be an easy way for people to introduce subtle bugs that make them not trust the build very much.

I'm not sure I understand this comment.  Is your concern that people will use it for tasks that they shouldn't?  If so, how is this different than any other normalizer in that regard?  For instance, how is this different than specifying relative path normalization when your task actually needs absolute path normalization?

Reply:
> Is your concern that people will use it for tasks that they shouldn't? If so, how is this different than any other normalizer in that regard?

Yes, this isn't really different. I was initially worried that there wasn't many other places where we could use this internally, but I think there are probably a few more places we could use this (e.g., config files for code quality tools) that I didn't think of yesterday.